File: 22.Automated-Testing-Unit-Integration-and-E2E/code/typescript/testing_http_handlers_apis_7.ts
import request from 'supertest';
import express from 'express';

const app = express();
app.use(express.json());
app.post('/api/v1/users', (req, res) => {
    res.status(201).json({ id: 1 });
});

describe('POST /api/v1/users', () => {
    it('returns 201', async () => {
        const response = await request(app)
            .post('/api/v1/users')
            .send({ email: 'a@x.com', name: 'Ada' })
            .set('Content-Type', 'application/json');

        expect(response.status).toBe(201);
        expect(response.headers['content-type']).toMatch(/application/json/);
    });
});

File: 22.Automated-Testing-Unit-Integration-and-E2E/code/rust/testing_http_handlers_apis_7.rs
use axum::{routing::post, Router, response::IntoResponse, http::StatusCode};
use axum::http::Request;
use hyper::Body;
use tower::ServiceExt;

async fn create_user() -> impl IntoResponse {
    (StatusCode::CREATED, "{"id": 1}")
}

#[tokio::test]
async fn test_create_user_returns_201() {
    let app = Router::new().route("/api/v1/users", post(create_user));
    
    let response = app
        .oneshot(Request::builder()
            .method("POST")
            .uri("/api/v1/users")
            .header("Content-Type", "application/json")
            .body(Body::from(r#"{"email":"a@x.com","name":"Ada"}"#))
            .unwrap())
        .await
        .unwrap();

    assert_eq!(response.status(), StatusCode::CREATED);
}
